azpubsub/contextvalidator/src/main/scala/kafka/security/auth/AzPubSubAclAuthorizer.scala

```diff
@@ -0,0 +1,459 @@
+    }
+  }
+
+  override def removeAcls(resource: Resource): Boolean = {
+    inWriteLock(lock) {
+      val result = zkClient.deleteResource(resource)
+      updateCache(resource, VersionedAcls(Set(), 0))
+      updateAclChangedFlag(resource)
+      result
+    }
+  }
+
+  override def getAcls(resource: Resource): Set[Acl] = {
+    inReadLock(lock) {
+      aclCache.get(resource).map(_.acls).getOrElse(Set.empty[Acl])
+    }
+  }
+
+  override def getAcls(principal: KafkaPrincipal): Map[Resource, Set[Acl]] = {
+    inReadLock(lock) {
+      aclCache.mapValues { versionedAcls =>
+        versionedAcls.acls.filter(_.principal == principal)
+      }.filter { case (_, acls) =>
+        acls.nonEmpty
+      }.toMap
+    }
+  }
+
+  override def getAcls(): Map[Resource, Set[Acl]] = {
+    inReadLock(lock) {
+      aclCache.mapValues(_.acls).toMap
+    }
+  }
+
+  def close() {
+    if (aclChangeListener != null) aclChangeListener.close()
+    if (zkClient != null) zkClient.close()
+  }
+
+
+  private[auth] def startZkChangeListeners(): Unit = {
+    aclChangeListener = new ZkNodeChangeNotificationListener(zkClient, AclChangeNotificationZNode.path, AclChangeNotificationSequenceZNode.SequenceNumberPrefix, AclChangedNotificationHandler)
+    aclChangeListener.init()
+  }
+
+  private def updateCache(resource: Resource, versionedAcls: VersionedAcls) {
+    if (versionedAcls.acls.nonEmpty) {
+      aclCache.put(resource, versionedAcls)
+    } else {
+      aclCache.remove(resource)
+    }
+  }
+
+  private def getAclsFromZk(resource: Resource): VersionedAcls = {
+    val acls = zkClient.getVersionedAclsForResource(resource)
+    VersionedAcls(acls.acls, acls.zkVersion)
+  }
+
+  private def loadCache()  {
+    inWriteLock(lock) {
+      val resourceTypes = zkClient.getResourceTypes()
+      for (rType <- resourceTypes) {
+        val resourceType = ResourceType.fromString(rType)
+        val resourceNames = zkClient.getResourceNames(resourceType.name)
+        for (resourceName <- resourceNames) {
+          val versionedAcls = getAclsFromZk(Resource(resourceType, resourceName))
+          updateCache(new Resource(resourceType, resourceName), versionedAcls)
+        }
+      }
+    }
+  }
+
+  private def updateAclChangedFlag(resource: Resource) {
+    zkClient.createAclChangeNotification(resource.toString)
+  }
+
+  /**
+    * Safely updates the resources ACLs by ensuring reads and writes respect the expected zookeeper version.
+    * Continues to retry until it successfully updates zookeeper.
+    *
+    * Returns a boolean indicating if the content of the ACLs was actually changed.
+    *
+    * @param resource the resource to change ACLs for
+    * @param getNewAcls function to transform existing acls to new ACLs
+    * @return boolean indicating if a change was made
+    */
+  private def updateResourceAcls(resource: Resource)(getNewAcls: Set[Acl] => Set[Acl]): Boolean = {
+    var currentVersionedAcls =
+      if (aclCache.contains(resource))
```

Comment:
This is a read lock

Reply:
You meant adding a lock? Actually the two calling methods (addAcls, removeAcls) of this method are already using a lock. Don't need to fix this.